lib/phoenix_live_view/lifecycle.ex

```diff
@@ -0,0 +1,217 @@
+
+  @lifecycle :lifecycle
+
+  @type t :: %__MODULE__{
+          handle_event: [Hook.t()],
+          handle_info: [Hook.t()],
+          handle_params: [Hook.t()],
+          mount: [Hook.t()]
+        }
+
+  defstruct handle_event: [], handle_info: [], handle_params: [], mount: []
+
+  def attach_hook(%Socket{} = socket, id, stage, fun)
+      when stage in [:handle_event, :handle_info, :handle_params] do
+    hook = Hook.new!(id, stage, fun)
+    lifecycle = lifecycle(socket)
+    existing = for h <- Map.fetch!(lifecycle, stage), h.id == id, do: h
+
+    unless existing == [] do
+      raise ArgumentError, """
+      existing hook #{inspect(hook.id)} already attached on #{inspect(hook.stage)}.
+      """
+    end
+
+    update_lifecycle(socket, stage, fn hooks ->
+      hooks ++ [hook]
+    end)
+  end
+
+  def attach_hook(%Socket{}, _id, stage, _fun) do
+    raise ArgumentError, """
+    invalid lifecycle event provided to attach_hook.
+
+    Expected one of: :handle_event | :handle_info | :handle_params
+
+    Got: #{inspect(stage)}
+    """
+  end
+
+  def detach_hook(%Socket{} = socket, id, stage)
+      when stage in [:handle_event, :handle_info, :handle_params] do
+    update_lifecycle(socket, stage, fn hooks ->
+      for hook <- hooks, hook.id != id, do: hook
+    end)
+  end
+
+  def detach_hook(%Socket{}, _id, stage) do
+    raise ArgumentError, """
+    invalid lifecycle event provided to detach_hook.
+
+    Expected one of: :handle_event | :handle_info | :handle_params
+
+    Got: #{inspect(stage)}
+    """
+  end
+
+  defp lifecycle(socket) do
+    Map.fetch!(socket.private, @lifecycle)
+  end
+
+  defp update_lifecycle(socket, stage, fun) do
+    lifecycle = lifecycle(socket)
+    new_lifecycle = Map.update!(lifecycle, stage, fun)
+    put_lifecycle(socket, new_lifecycle)
+  end
+
+  defp put_lifecycle(socket, lifecycle) do
+    put_private(socket, @lifecycle, lifecycle)
+  end
+
+  defp put_private(%Socket{private: private} = socket, key, value) when is_atom(key) do
+    %{socket | private: Map.put(private, key, value)}
+  end
+
+  # Lifecycle Event API
+
+  @doc false
+  def on_mount(view, view), do: raise_own_mount_hook!(view, view)
+  def on_mount(view, {view, :mount} = id), do: raise_own_mount_hook!(view, id)
+
+  def on_mount(_view, {module, fun} = id) when is_atom(module) and is_atom(fun) do
+    Hook.new!(id, :mount, Function.capture(module, fun, 3))
+  end
+
+  def on_mount(_view, module) when is_atom(module) do
+    Hook.new!(module, :mount, Function.capture(module, :mount, 3))
+  end
+
+  def on_mount(view, result) do
+    raise ArgumentError, """
+    invalid on_mount hook declared in #{inspect(view)}.
+
+    Expected one of:
+
+        Module
+        {Module, Function}
+
+    Got: #{inspect(result)}
+    """
+  end
+
+  @doc false
+  def mount(_view, hooks) when is_list(hooks) do
+    Enum.reduce(hooks, %__MODULE__{}, fn %Hook{} = hook, acc ->
+      %{acc | mount: [hook | acc.mount]}
+    end)
```

Comment:
```suggestion
    %__MODULE__{mount: Enum.reverse(hooks)}
```

?

Reply:
lol yes there is nothing to append to at this point :)